hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/StakingRewardsHandlerImpl.java

```diff
@@ -0,0 +1,375 @@
+            //   was already rewarded earlier today.
+            //   3. Alice's stakePeriodStart is the first non-rewardable period, but she was not
+            //   rewarded today.
+            //   4. Alice's stakePeriodStart is the current period.
+            // We need to record her current stake as totalStakeAtStartOfLastRewardedPeriod in
+            // scenarios 1 and 3, but not 2 and 4. (As noted below, in scenario 2 we want to
+            // preserve an already-recorded memory of her stake at the beginning of this period;
+            // while in scenario 4 there is no point in recording anything---it will go unused.)
+            if (earnedZeroRewardsBecauseOfZeroStake(account, stakingRewardStore, consensusNow)) {
+                return true;
+            }
+            if (account.stakeAtStartOfLastRewardedPeriod() != NOT_REWARDED_SINCE_LAST_STAKING_META_CHANGE) {
+                // Alice was in a reward situation, but did not earn anything because she already
+                // received a reward earlier today; we preserve our memory of her stake from then
+                return false;
+            }
+            // Alice was in a reward situation for the first time today, but received nothing---
+            // either because she is staking < 1 hbar, or because she started staking only
+            // today or yesterday; we don't care about the exact reason, we just remember
+            // her total stake as long as she didn't begin staking today exactly
+            return account.stakePeriodStart() < stakePeriodManager.currentStakePeriod(consensusNow);
+        }
+    }
+
+    private long rewardableStakeStartFor(
+            final boolean rewardsActivated, @NonNull final Account account, @NonNull final Instant consensusNow) {
+        if (!rewardsActivated || account.declineReward()) {
+            return 0;
+        }
+        final var startPeriod = account.stakePeriodStart();
+        final var currentPeriod = stakePeriodManager.currentStakePeriod(consensusNow);
+        if (startPeriod >= currentPeriod) {
+            return 0;
+        } else {
+            if (isRewardedSinceLastStakeMetaChange(account) && (startPeriod == currentPeriod - 1)) {
+                // Special case---this account was already rewarded today, so its current stake
+                // has almost certainly changed from what it was at the start of the day
+                return account.stakeAtStartOfLastRewardedPeriod();
+            } else {
+                return roundedToHbar(totalStake(account));
+            }
+        }
+    }
+
+    private void decreaseStakeRewardAccountBalance(
+            final Map<AccountID, Long> rewardsPaid,
+            final AccountID stakingRewardAccountId,
+            final WritableAccountStore writableAccountStore) {
+        if (!rewardsPaid.isEmpty()) {
+            final var totalPaidRewards =
+                    rewardsPaid.values().stream().mapToLong(Long::longValue).sum();
```

Comment:
I'm confused about when using `.stream()` is allowed and when it isn't...?

Reply:
Removed